### PR TITLE
Enable FormattingAnalyzer

### PR DESCRIPTION
**Pull request by sharwell**
Recommend reviewing by individual commit.

Reviewers should pay particular attention to all commits that are not the large and automated application of formatting changes.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder.cs [hunk 1]
```diff
@@ -687,7 +687,7 @@ internal BoundStatement WrapWithVariablesIfAny(CSharpSyntaxNode scopeDesignator,
             }
 
             return new BoundBlock(statement.Syntax, locals, ImmutableArray.Create(statement))
-                { WasCompilerGenerated = true };
+            { WasCompilerGenerated = true };
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder.cs]: This may be a degradation.

File: src/Compilers/CSharp/Portable/Binder/ForLoopBinder.cs [hunk 2]
```diff
@@ -90,12 +90,12 @@ private BoundForStatement BindForParts(ForStatementSyntax node, Binder originalB
...
                     {
                         increment = new BoundBlock(scopeDesignator, locals, ((BoundStatementList)increment).Statements)
-                                            { WasCompilerGenerated = true };
+                        { WasCompilerGenerated = true };
                     }
                     else
                     {
                         increment = new BoundBlock(increment.Syntax, locals, ImmutableArray.Create(increment))
-                                            { WasCompilerGenerated = true };
+                        { WasCompilerGenerated = true };
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/ForLoopBinder.cs]: another case.   i feel lke indentation should be allowed at certain depths.  either a depth of '1', or 'matching with 'new' should be allowed.

File: src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs [hunk 3]
```diff
@@ -324,7 +324,7 @@ internal virtual NamedTypeSymbol LookupMetadataType(ref MetadataTypeName emitted
                 }
             }
 
-        Done:
+Done:
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs]: i believe the former is a supported label position.  perhaps the compiler/Roslyn should be setting that formatting option in editor config?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs [hunk 4]
```diff
@@ -821,9 +821,9 @@ internal override void ForceComplete(SourceLocation locationOpt, CancellationTok
                 state.SpinWaitComplete(incompletePart, cancellationToken);
             }
 
-        done:
-            // Don't return until we've seen all of the CompletionParts. This ensures all
-            // diagnostics have been reported (not necessarily on this thread).
+done:
+// Don't return until we've seen all of the CompletionParts. This ensures all
+// diagnostics have been reported (not necessarily on this thread).
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs]: likely formatting bug.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs [hunk 5]
```diff
@@ -1122,7 +1122,7 @@ static S1 MayWrap(ref int arg)
     }
 ";
             CreateCompilationWithMscorlibAndSpan(text).VerifyDiagnostics(
-                // no diagnostics expected
+            // no diagnostics expected
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs]: this change is minor, but seems also undesirable.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs]: probably not a big deal though.

File: src/Compilers/Core/Portable/Operations/CommonConversion.cs [hunk 6]
```diff
@@ -16,11 +16,11 @@ public struct CommonConversion
...
-            Exists =        1 << 0,
-            IsIdentity =    1 << 1,
-            IsNumeric =     1 << 2,
-            IsReference =   1 << 3,
-            IsImplicit =    1 << 4,
+            Exists = 1 << 0,
+            IsIdentity = 1 << 1,
+            IsNumeric = 1 << 2,
+            IsReference = 1 << 3,
+            IsImplicit = 1 << 4,
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Operations/CommonConversion.cs]: Seems worse.

**Comment by CyrusNajmabadi** [comment]
PR hard to review at this size.  Talked with sam offline.  Let's break out totally safe changes (like removing trailing whitespace).  That should greatly decrease the size of the diff.  Then we can do successful commits that are easier to review.

File: src/Compilers/Core/Portable/InternalUtilities/TextKeyedCache.cs [hunk 7]
```diff
@@ -51,13 +51,17 @@ public SharedEntryValue(string Text, T item)
         // local cache
         // simple fast and not threadsafe cache 
         // with limited size and "last add wins" expiration policy
+#pragma warning disable IDE0055 // Fix formatting
         private readonly (string Text, int HashCode, T Item)[] _localTable = new (string Text, int HashCode, T Item)[LocalSize];
+#pragma warning restore IDE0055 // Fix formatting
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/InternalUtilities/TextKeyedCache.cs]: What's up with formatting here?

File: src/Compilers/CSharp/Portable/Binder/Binder.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder.cs]: It's not typical to indent the open brace for an object initializer, is it?

File: src/Compilers/CSharp/Portable/Binder/ForLoopBinder.cs (hunk 2 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Binder/ForLoopBinder.cs]: ➡️ Duplicate of https://github.com/dotnet/roslyn/pull/31142#discussion_r233238801

File: src/Compilers/Core/Portable/Operations/CommonConversion.cs (hunk 6 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/Operations/CommonConversion.cs]: ➡️ By design. Our .editorconfig is not configured to ignore these spaces.

File: src/Compilers/Core/Portable/InternalUtilities/TextKeyedCache.cs (hunk 7 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/InternalUtilities/TextKeyedCache.cs]: Command line builds run with the 15.7 formatter. The 15.8 formatter includes a bug fix for this syntax (space after `new`). There is no way to make both the analyzer and the IDE happy, so the 15.8 style is preferred and the command line warning is suppressed.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs [hunk 8]
```diff
@@ -981,7 +982,7 @@ private static BoundExpression MakeLiteral(SyntaxNode syntax, ConstantValue cons
 
             if (storesToTemps.Count > 0)
             {
-                int tempsNeeded = MergeArgumentsAndSideEffects(arguments,storesToTemps);
+                int tempsNeeded = MergeArgumentsAndSideEffects(arguments, storesToTemps);
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs]: I extracted indentation changes, and then I extracted wrapping changes. The resulting "Format documents in solution" change should be left with whitespace changes only.

File: src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexLexer.cs [hunk 9]
```diff
@@ -364,12 +364,18 @@ private bool IsOptionChar(char ch)
...
+                case 'i':
+                case 'I':
+                case 'm':
+                case 'M':
+                case 'n':
+                case 'N':
+                case 's':
+                case 'S':
+                case 'x':
+                case 'X':
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexLexer.cs]: :'(

Note that the original formatting was intentional.  Though i can live with this.

**Comment by CyrusNajmabadi** [comment]
Went through all the change.  LGTM.  Thanks for breaking it out into pieces.  It made reviewing much easier!

File: src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexLexer.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexLexer.cs]: Do we not have: 

image

On?  I would consider this a version of that rule.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexLexer.cs]: Note: i'm still ok with this.  I prefer the previous code.  But i wouldn't block the formatting analyzer on this.

File: src/Compilers/Core/Portable/InternalUtilities/TextKeyedCache.cs [hunk 10]
```diff
@@ -51,13 +51,17 @@ public SharedEntryValue(string Text, T item)
         // local cache
         // simple fast and not threadsafe cache 
         // with limited size and "last add wins" expiration policy
+#pragma warning disable IDE0055 // Fix formatting
```
Comment (sharwell) [inline src/Compilers/Core/Portable/InternalUtilities/TextKeyedCache.cs]: 📝 This will be fixed by #31276

Comment (sharwell) [reply src/Compilers/Core/Portable/InternalUtilities/TextKeyedCache.cs]: ➡️ Now fixed and incorporated.

Comment (sharwell) [reply src/Compilers/Core/Portable/InternalUtilities/TextKeyedCache.cs]: ➡️ Now fixed and updated

**Comment by jaredpar** [comment]
We should separate out the actual formatting change from enabling the check in CI. It's difficult to review the changes together.

**Comment by sharwell** [comment]
@jaredpar @tmat The only substantial difference in the last force-push is I rewrote daf181f12ccfcdd62fe94896d963a493eae04481 to account for the switch to arcade.

File: build/scripts/tools.ps1 [hunk 11]
```diff
@@ -413,7 +413,7 @@ function MSBuild() {
   $cmdArgs = "$($buildTool.Command) /m /nologo /clp:Summary /v:$verbosity /nr:$nodeReuse"
 
   if ($warnAsError) { 
-    $cmdArgs += " /warnaserror /p:TreatWarningsAsErrors=true" 
+    $cmdArgs += " /p:TreatWarningsAsErrors=true"
```
Comment (sharwell) [inline build/scripts/tools.ps1]: @tmat I had to make this change or the build would ignore my request for `<WarningsNotAsErrors>`.

File: build/Targets/Settings.props [hunk 12]
```diff
@@ -178,6 +178,11 @@
     </When>
   </Choose>
 
+  <PropertyGroup>
+    <!-- Don't treat FormattingAnalyzer as an error, even when TreatWarningsAsErrors is specified. -->
+    <WarningsNotAsErrors>$(WarningsNotAsErrors);IDE0055</WarningsNotAsErrors>
```
Comment (jaredpar) [inline build/Targets/Settings.props]: This seems like it will cause IDE055 to warn on every CI leg instead of just the correctness leg.

**Comment by sharwell** [comment]
> Need to disable the analyzer warnings on only the build correctness leg.

@jaredpar Since this now treats warnings as warnings, this analyzer will not block _any_ leg. I suggest we leave this as a follow-up item since it's a non-breaking visual artifact.

**Comment by jaredpar** [comment]
> I suggest we leave this as a follow-up item since it's a non-breaking visual artifact. 

This is a fairly simple change to make. I don't see any reason to delay doing this.

File: src/Compilers/CSharp/Portable/Binder/Binder.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Binder/Binder.cs]: This seems like the desired behavior to me. If someone disagrees, please feel free to create an issue. 😄

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs [hunk 13]
```diff
@@ -477,13 +477,14 @@ private BoundExpression MakeConversionNode(BoundExpression rewrittenOperand, Typ
...
-                            conversion,             
+                            conversion,
                             @checked: @checked,
                             explicitCastInCode: false,
                             conversionGroupOpt: null,
                             constantValueOpt: default(ConstantValue),
                             type: type,
-                            hasErrors: !conversion.IsValid) { WasCompilerGenerated = true };
+                            hasErrors: !conversion.IsValid)
+            { WasCompilerGenerated = true };
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs]: This seems like the desired behavior to me. If someone disagrees, please feel free to create an issue. 😄

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs (hunk 5 above)
Comment (sharwell) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs]: I agree this appears to be a bug.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs (hunk 4 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs]: I agree, this appears to be a bug.

**Comment by jaredpar** [comment]
Pretty sure this commit should do the trick 

https://github.com/jaredpar/roslyn/commit/c621d527db34eb49c2facaf082ebd78a67b850b2

**Comment by sharwell** [comment]
@jinujoseph for approval

**Pull request merged**